### PR TITLE
auto-pts enhancements - many auto-pts instances on one machine

### LINKED ISSUE #6: Evaluate running two auto-pts instances on one machine
Investigate effort to allow running two instances on one virtual machine. 
- Add option to specify autopts server port or automatically select a free port

### PR DESCRIPTION
**Pull request by mkasenberg**
Main new features:
- many autoptsserver instances on one Windows machine
- many autoptsclient-zephyr instances on one Windows machine or Linux host
- one autoptsclient_bot on Windows machine

Note: for autoptsclient-zephyr running on Windows, download socat.exe from https://sourceforge.net/projects/unix-utils/files/socat/1.7.3.2/ and add its directory to PATH.

Example runs:

Run many instances of autoptsserver on one Windows machine like:
- in first console:
$ python autoptsserver.py
- in second console:
$ python autoptsserver.py -S 65002
- and more.

Run many autoptsclient-zephyr.py instances on one machine, for Windows it can looks like:
- in third console: (-S 65000 -C 65001 by default)
$ python ./autoptsclient-zephyr.py zephyr-master ~/zephyrproject/zephyr/tests/bluetooth/tester/build/zephyr/zephyr.elf -t COM3 -b nrf52
- in fourth console:
$ python ./autoptsclient-zephyr.py zephyr-master ~/zephyrproject/zephyr/tests/bluetooth/tester/build/zephyr/zephyr.elf -t COM4 -b nrf52 -S 65002 -C 65003
- and more.

On Linux it can looks like:
- in first console:
$ ./autoptsclient-zephyr.py zephyr-master ~/zephyrproject/zephyr/tests/bluetooth/tester/build/zephyr/zephyr.elf -i 192.168.4.2 192.168.4.2 -l 192.168.4.1 192.168.4.1 -t /dev/ttyACM0 -b nrf52
- in second console:
$ ./autoptsclient-zephyr.py zephyr-master ~/zephyrproject/zephyr/tests/bluetooth/tester/build/zephyr/zephyr.elf -i 192.168.4.2 192.168.4.2 -l 192.168.4.1 192.168.4.1 -t /dev/ttyACM1 -b nrf52  -S 65002 -C 65003

Example run of autoptsclient-zephyr for MESH testing:
- for Windows, in third console:
$ python ./autoptsclient-zephyr.py zephyr-master ~/zephyrproject/zephyr/tests/bluetooth/tester/build/zephyr/zephyr.elf -t COM3 -b nrf52 -c MESH -S 65000 65002 -C 65001 65003
- or for Linux host:
$ ./autoptsclient-zephyr.py zephyr-master ~/zephyrproject/zephyr/tests/bluetooth/tester/build/zephyr/zephyr.elf -i 192.168.4.2 192.168.4.2 -l 192.168.4.1 192.168.4.1 -t /dev/ttyACM0 -b nrf52  -c MESH -S 65000 65002 -C 65001 65003

Config for bot has new elements:
z['auto_pts'] = {
    # 'server_ip': ['192.168.3.2', '192.168.3.2'],  # uncomment for Linux host
    # 'local_ip' : ['192.168.3.1', '192.168.3.1'],  # uncomment for Linux host
    'cli_port': [65001, 65003],
    'srv_port': [65000, 65002],
    'project_path': 'C:/Users/Magda/zephyrproject/zephyr',
    'workspace': 'zephyr-master',
    'board': 'nrf52',
    'enable_max_logs': False,
    'retry': 0,
    'bd_addr': '',
}

### CODE REVIEW

File: autoptsclient_common.py [hunk 1]
```diff
@@ -957,3 +957,9 @@ def __init__(self, description):
         self.add_argument("-r", "--retry", type=int, default=0,
                           help="Repeat test if failed. Parameter specifies "
                                "maximum repeat count per test")
+
+        self.add_argument("-S", "--srv_port", type=int, default=SERVER_PORT,
+                          help="Specify the server port number")
+
+        self.add_argument("-C", "--cl_port", type=int, default=CLIENT_PORT,
```
Comment (michal-narajowski) [inline autoptsclient_common.py:964]: cl_port -> cli_port to match srv_port

File: autoptsclient-zephyr.py [hunk 2]
```diff
@@ -96,7 +101,13 @@ def parse_args():
 
 def main():
     """Main."""
-    if os.geteuid() == 0:  # root privileges are not needed
+
+    if sys.platform == "win32":
```
Comment (MariuszSkamra) [inline autoptsclient-zephyr.py:105]: you could use the 'have_admin_rights' function from winutils.py

File: ptsprojects/zephyr/iutctl.py [hunk 3]
```diff
@@ -62,6 +62,12 @@ def __init__(self, kernel_image, tty_file, board_name=None, use_rtt2pty=None):
         self.kernel_image = kernel_image
         self.tty_file = tty_file
 
+        get_debugger_snr_cmd = 'nrfjprog --com | awk \'/' + tty_file.replace('/', '\\/') + '/ { print $1 }\''
```
[the remark this answers is not in the document]
Comment (mkasenberg) [reply ptsprojects/zephyr/iutctl.py]: Good point. Moreover, after rebase with Windows workaround commit, I forgot to add the Windows version :)

File: bot/zephyr.py [hunk 4]
```diff
@@ -212,20 +231,31 @@ class PtsInitArgs(object):
...
         self.bd_addr = args["bd_addr"]
         self.enable_max_logs = args["enable_max_logs"]
         self.retry = args["retry"]
         self.test_cases = []
         self.excluded = []
+        self.srv_port = args["srv_port"]
+        self.cli_port = args["cli_port"]
+
+        try:
+            self.ip_addr = args['server_ip']
```
Comment (michal-narajowski) [inline bot/zephyr.py:244]: `if 'server_ip' in args:`

File: bot/zephyr.py [hunk 5]
```diff
@@ -212,20 +231,31 @@ class PtsInitArgs(object):
...
+            self.local_addr = args['local_ip']
+        except Exception as e:
+            if sys.platform == 'win32':
```
Comment (michal-narajowski) [inline bot/zephyr.py:247]: Use localhost as default address regardless of platform.

File: bot/common.py [hunk 6]
```diff
@@ -452,6 +455,44 @@ def update_repos(project_path, git_config):
...
+    else:
+        reg = "[0-9]+\s+/dev/\w+(?=\s+.+)"
+
+    try:
+        debuggers = re.findall(reg, debuggers)
+        for d in debuggers:
+            srn = d.split()[0]
+            dev = d.split()[1]
+
+            if not devices_in_use.__contains__(srn):
```
Comment (michal-narajowski) [inline bot/common.py:477]: `if not srn in devices_in_use`:

File: bot/common.py [hunk 7]
```diff
@@ -452,6 +455,44 @@ def update_repos(project_path, git_config):
...
+                devices_in_use.append(srn)
+                tty = dev
+                jlink = srn
+                break
+    except:
+        raise Exception("No jlinks found")
```
Comment (michal-narajowski) [inline bot/common.py:483]: I would prefer handling errors when they appear instead of catching everything with a try/except.

**Review by michal-narajowski — CHANGES_REQUESTED** [review]
Looks great! Just a few minor comments.

**Review by michal-narajowski — APPROVED** [review]
LGTM

**Pull request merged**